cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -0,0 +1,224 @@
+    }
+  }
+
+  /**
+   * Initializes the Git repository by cloning remote.
+   *
+   * @throws GitAPIException               when a Git operation fails.
+   * @throws AuthenticationConfigException when correct authentication credentials are not available.
+   * @throws IOException                   when file or network I/O fails.
+   */
+  @Override
+  protected void startUp() throws Exception {
+    if (git != null) {
+      return;
+    }
+    // Clean up the directory if it already exists.
+    deletePathIfExists(sourceControlConfig.getLocalRepoPath());
+    RepositoryConfig repositoryConfig = sourceControlConfig.getRepositoryConfig();
+    CloneCommand command = createCommand(Git::cloneRepository).setURI(repositoryConfig.getLink())
+                                                              .setDirectory(sourceControlConfig.getLocalRepoPath()
+                                                                                               .toFile());
+
+    String branch = getBranchRefName(repositoryConfig.getDefaultBranch());
+    if (branch != null) {
+      command.setBranchesToClone(Collections.singleton((branch))).setBranch(branch);
+    }
+    git = command.call();
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (git != null) {
+      git.close();
+    }
+    git = null;
+    deletePathIfExists(sourceControlConfig.getReposBasePath());
+  }
+
+  private static void doValidate(SecureStore secureStore, SourceControlConfig sourceControlConfig) throws IOException {
+    // Validate existence of secret in Secure Store.
+    RepositoryConfig config = sourceControlConfig.getRepositoryConfig();
+    LOG.info("Checking existence of secret in secure store: {}", config.getAuth().getTokenName());
+    try {
+      secureStore.get(sourceControlConfig.getNamespaceID(), config.getAuth().getTokenName()).get();
+    } catch (Exception e) {
+      throw new RepositoryConfigValidationException("Failed to get auth credentials from secure store.", e);
+    }
+
+    // Try fetching heads in the remote repository.
+    FileRepository localRepo;
+    Path validationRepoPath = sourceControlConfig.getValidationRepoPath();
+    deletePathIfExists(validationRepoPath);
+    try {
+      localRepo = new FileRepository(validationRepoPath.toFile());
+    } catch (IOException e) {
+      throw new IOException(String.format("Failed to create local git repository at path '%s'", validationRepoPath), e);
+    }
+
+    LOG.info("Created local repo in directory for validation: {}", validationRepoPath);
+    try (Git git = new Git(localRepo)) {
+      LsRemoteCommand cmd = createCommand(git::lsRemote, sourceControlConfig, secureStore).setRemote(config.getLink())
+                                                                                          .setHeads(true)
+                                                                                          .setTags(false);
+      validateDefaultBranch(cmd.call(), config.getDefaultBranch());
```

Comment:
Use `callAsMap` instead, then the branch name is in the Map key, which make it easier to check if the default branch exists.

Reply:
Thanks for the tip, changed. Wasn't aware of this method.